plugin/src/main/java/io/jenkins/plugins/redis/RedisFingerprintStorage.java

```diff
@@ -0,0 +1,189 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package io.jenkins.plugins.redis;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import hudson.BulkChange;
+import hudson.Extension;
+import hudson.Util;
+import hudson.model.FingerprintStorage;
+import hudson.model.Job;
+import hudson.model.Fingerprint;
+import hudson.util.PersistedList;
+
+import java.io.IOException;
+import java.io.ByteArrayInputStream;
+import java.io.StringWriter;
+import java.io.InputStream;
+import java.util.logging.Logger;
+import java.util.logging.Level;
+
+import jenkins.model.FingerprintFacet;
+import jenkins.util.SystemProperties;
+
+import org.jenkinsci.main.modules.instance_identity.InstanceIdentity;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import redis.clients.jedis.Jedis;
+
+/**
+ * Pluggable external fingerprint storage for fingerprints in Redis.
+ */
+
+@Extension
+public class RedisFingerprintStorage extends FingerprintStorage {
+
+    private final String instanceId;
+    private static final Logger logger = Logger.getLogger(Fingerprint.class.getName());
+
+    @Restricted(NoExternalUse.class)
+    private static final String host = SystemProperties.getString("Redis.host", "localhost");
+    @Restricted(NoExternalUse.class)
+    private static final Integer port = SystemProperties.getInteger("Redis.port", 6379);
+
+    /**
+     * Saves the given fingerprint.
+     */
+    public RedisFingerprintStorage() throws IOException{
+        try {
+            instanceId = Util.getDigestOf(new ByteArrayInputStream(InstanceIdentity.get().getPublic().getEncoded()));
+        } catch (IOException e) {
+            logger.log(Level.WARNING, "Failed to obtain Instance ID. "+e);
+            throw e;
+        }
+    }
+
+    /**
+     * Creates a connection to the database.
+     */
+    private Jedis getJedis() {
+        return new Jedis(host, port);
+    }
+
+    /**
+     * Saves the given fingerprint.
+     */
+    public synchronized void save(Fingerprint fp) throws IOException {
+        if(BulkChange.contains(fp))
```

Comment:
either single-line this or add outer brackets https://blog.codecentric.de/en/2014/02/curly-braces/

Reply:
Thanks @afalko Nice read :) I have now incorporated this inside the core PR at https://github.com/jenkinsci/jenkins/pull/4731/, because I have abstracted this part away from plugin implementation.